### PR TITLE
Add US (Southwest Power Pool)

### PR DESCRIPTION
**Pull request by systemcatch**
The production parser is working nicely. However the exchange data provided by SPP sometimes differs from the MISO scheduled data by a significant amount. I've therefore left the US-MISO->US-SPP exchange disabled until this discrepancy can be resolved (function is all built). Practically we might be best waiting until the MISO api provides actual exchange data. :crossed_fingers: 

/cc @tmslaine 

ref #1177 

spp

### CODE REVIEW

File: parsers/US_SPP.py [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+    Takes a dataframe and logging instance as input.
+    Checks for new generation types and logs awarning if any are found.
+    Parses the dataframe row by row removing unneeded keys.
+    Returns a list of 2 element tuples, each containing a datetime object
+    and production dictionary.
+    """
+
+    # Remove leading whitespace in column headers.
+    df.columns = df.columns.str.strip()
+    print(df)
```
Comment (corradio) [inline parsers/US_SPP.py]: Maybe remove this?

Comment (systemcatch) [reply parsers/US_SPP.py]: :astonished: whoops that was left over from debugging.

File: parsers/US_SPP.py [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+import pandas as pd
+import requests
+
+generation_url = 'https://marketplace.spp.org/public-data-api/gen-mix/asFile'
+
+exchange_url = 'https://marketplace.spp.org/public-data-api/interchange-trend/asFile'
+
+mapping = {'coal': 'coal',
+           'unknown': 'unknown',
+           'Wind': 'wind',
```
Comment (jarek) [inline parsers/US_SPP.py]: Just checking - I can't load the source files, but do they really have "coal" lowercase and most everything else uppercase?

File: parsers/US_SPP.py [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+           'Nuclear': 'nuclear',
+           'Hydro': 'hydro',
+           'Solar': 'solar',
+           'Natural Gas': 'gas',
+           'Diesel Fuel Oil': 'oil',
+           'Waste Disposal Services': 'biomass'
+            }
+
+tie_mapping = {'US-MISO->US-SPP': ['AMRN', 'DPC', 'GRE', 'MDU', 'MEC', 'NSP', 'OTP']}
```
Comment (jarek) [inline parsers/US_SPP.py]: I think pylint's preferred style is to have the constants outside the function ALL_CAPS. This actually caught me out when I was reading data_processor() and was wondering where `mapping` came from.

File: parsers/US_SPP.py [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+
+    # Remove leading whitespace in column headers.
+    df.columns = df.columns.str.strip()
+
+    keys_to_remove = ['Coal Market', 'Coal Self', 'GMT MKT Interval', 'Average Actual Load',
+                      'Other', 'Waste Heat']
+
+    # Check for new generation columns.
+    known_keys = list(mapping.keys()) + keys_to_remove
+    column_headers = list(df)
```
Comment (jarek) [inline parsers/US_SPP.py]: Would `list(df.columns)` also work here? I think that's a bit clearer in terms of what it is

File: parsers/US_SPP.py [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+
+        extra_unknowns = sum([production[k] for k in unknown_keys])
+        production['unknown'] = production['Other'] + production['Waste Heat'] + extra_unknowns
+
+        dt_aware = parser.parse(production['GMT MKT Interval'])
+
+        for k in keys_to_remove:
+            production.pop(k, None)
+
+        mapped_production = {mapping.get(k,k):v for k,v in production.items()}
```
Comment (jarek) [inline parsers/US_SPP.py]: I think at this point we've assured that `production` only contains keys that are in `mapping`, so the default value in `mapping.get(k, k)` is not needed?

File: parsers/US_SPP.py [hunk 6]
```diff
@@ -0,0 +1,198 @@
...
+    if target_datetime:
+        raise NotImplementedError('This parser is not yet able to parse past dates')
+
+    raw_data = get_data(exchange_url, session=session)
+    sorted_codes = '->'.join(sorted([zone_key1, zone_key2]))
+
+    try:
+        exchange_ties = tie_mapping[sorted_codes]
+    except KeyError as e:
+        raise NotImplementedError('This exchange pair is not implemented')
```
Comment (jarek) [inline parsers/US_SPP.py]: Maybe include `sorted_codes` in the error message for possible later debugging

File: parsers/US_SPP.py [hunk 7]
```diff
@@ -0,0 +1,198 @@
+#!usr/bin/env python3
+
+"""Parser for the Southwest Power Pool area of the United States."""
+
+from dateutil import parser, tz
```
Comment (jarek) [inline parsers/US_SPP.py]: It looks like `tz` is not actually used?

File: parsers/US_SPP.py [hunk 8]
```diff
@@ -0,0 +1,198 @@
...
+    column_headers = list(df)
+
+    unknown_keys = []
+    for heading in column_headers:
+        if heading not in known_keys:
+            logger.warning('New column \'{}\' present in US-SPP data source.'.format(
+                heading), extra={'key': 'US-SPP'})
+            unknown_keys.append(heading)
+
+    keys_to_remove += unknown_keys
```
Comment (jarek) [inline parsers/US_SPP.py]: Completely optional, the above switched to using sets:

```
keys_to_remove = {'Coal Market', 'Coal Self', 'GMT MKT Interval', 'Average Actual Load',
    'Other', 'Waste Heat'}  # set literal
known_keys = mapping.keys() | keys_to_remove  # this is a set union
column_headers = set(df.columns)  # check exact syntax

unknown_keys = column_headers - known_keys  # set difference

for heading in unknown_keys:
    logger.warning('New column \'{}\' present in US-SPP data source.'.format(heading),
        extra={'key': 'US-SPP'})

keys_to_remove = keys_to_remove | unknown_keys  # can also do "|=" but it's a bit unclear I feel
```

Untested! But later on in the code, sets can be iterated over same as lists.

File: parsers/US_SPP.py (hunk 2 above)
Comment (systemcatch) [reply parsers/US_SPP.py]: Nope coal and the unknown key are leftovers that can be removed.

File: parsers/US_SPP.py (hunk 4 above)
Comment (systemcatch) [reply parsers/US_SPP.py]: It works and is clearer, swapping to that.

File: parsers/US_SPP.py (hunk 8 above)
Comment (systemcatch) [reply parsers/US_SPP.py]: Would you say there is a benefit to using sets here rather than lists?

File: parsers/US_SPP.py (hunk 5 above)
Comment (systemcatch) [reply parsers/US_SPP.py]: With the 'coal' and 'unknown' keys gone from MAPPING, get stops them raising KeyErrors in the dict comprehension. So I think it needs to stay.

File: parsers/US_SPP.py (hunk 8 above)
Comment (maxbellec) [reply parsers/US_SPP.py]: it's certainly not necessary, but makes computing `unknown_keys` easier and clearer

**Comment by systemcatch** [comment]
All good to go here guys.

**Comment by alixunderplatz** [comment]
I have some doubts that only US-NE, US-KS and US-OK are within the borders. Seems like North and South Dakota could be withdrawn from MISO and added to SWPP as well.

image

image

image

image

**Comment by alixunderplatz** [comment]
I hope one day all the US-RTOs will exchange their assets with each other to clean this mess up and create coherent areas.

This is from SPPs front page:

image

**Comment by systemcatch** [comment]
It's very confusing. I've seen maps showing the Dakotas as part of MISO. Maybe we compromise and allocate South Dakota to SPP for now?

**Comment by alixunderplatz** [comment]
@systemcatch I'd say MISO has rather a small influence on North Dakota. I also think the first two maps in my comment are the most recent ones and they don't show any MISO in ND anymore. So SPP covers most of the area of ND now.

Also, looking at population density, most of the consumption in MISO will come from areas in its "core" and not from ND.
image

So I'd also push North Dakota over to SPP.

**Comment by systemcatch** [comment]
Good points I'll move them both over.

**Comment by systemcatch** [comment]
All done I'll merge if tests pass.

dakota

**Comment by Hyric89** [comment]
Capacity fleet by source as done in #1177 

Wind: 16.13 GW
Hydro: 3.42 GW
Nuclear: 2.11 GW
Gas: 34.26 GW
Coal: 26.06 GW
Oil: 1.69 GW
Other: 0.29 GW

**Pull request merged**